storages/jsonl-storage/src/lib.rs

```diff
@@ -0,0 +1,396 @@
+    }
+
+    async fn scan_data(&self, table_name: &str) -> Result<RowIter> {
+        self.scan_data(table_name)
+    }
+}
+
+fn read_lines<P>(filename: P) -> io::Result<io::Lines<io::BufReader<File>>>
+where
+    P: AsRef<Path>,
+{
+    let file = File::open(filename)?;
+    Ok(io::BufReader::new(file).lines())
+}
+
+#[async_trait(?Send)]
+impl StoreMut for JsonlStorage {
+    async fn insert_schema(&mut self, schema: &Schema) -> Result<()> {
+        let path = format!("{}/{}.json", self.path.display(), schema.table_name);
+        let path = PathBuf::from(path);
+
+        if schema.column_defs.is_some() {
+            self.write_schema(schema)?
+        }
+
+        File::create(path).map_storage_err()?;
+        JsonlStorage::insert_schema(self, schema);
+
+        Ok(())
+    }
+
+    async fn delete_schema(&mut self, table_name: &str) -> Result<()> {
+        if let Ok(table_path) = JsonlStorage::data_path(self, table_name) {
+            let schema_path = JsonlStorage::schema_path(self, table_name);
+
+            remove_file(table_path).map_storage_err()?;
+            if let Some(schema_path) = schema_path {
+                remove_file(schema_path).map_storage_err()?;
+            }
+
+            JsonlStorage::delete_schema(self, table_name);
+        }
+
+        Ok(())
+    }
+
+    async fn append_data(&mut self, table_name: &str, rows: Vec<DataRow>) -> Result<()> {
+        self.tables
+            .get(table_name)
+            .map_storage_err(JsonlStorageError::TableDoesNotExist.to_string())
+            .and_then(|schema| {
+                let table_path = JsonlStorage::data_path(self, table_name)?;
+
+                let mut file = OpenOptions::new()
+                    .write(true)
+                    .append(true)
+                    .open(table_path)
+                    .map_storage_err()?;
+
+                for row in rows {
+                    let json_string = match row {
+                        DataRow::Map(hash_map) => {
+                            hash_map
+                                .into_iter()
+                                // todo! why even schemaless get to here? on_where.rs L55
+                                .map(|(key, value)| {
+                                    let value = JsonValue::try_from(value)?.to_string();
+
+                                    Ok(format!("\"{key}\": {value}"))
+                                })
+                                .collect::<Result<Vec<_>>>()?
+                        }
+                        DataRow::Vec(values) => {
+                            match &schema.column_defs {
+                                Some(column_defs) => {
+                                    // todo! validate columns?
+                                    column_defs
+                                        .iter()
+                                        .map(|column_def| column_def.name.clone())
+                                        .zip(values.into_iter())
```

Comment:
Is there a way to merge two different Iterators? like..
```rs
// 🦀 This code does not compile!
let json_iter: Box<dyn Iterator> = match row {
  DataRow::Map(hash_map) => hash_map.into_iter(),
  DataRow::Vec(values) => match &schema.column_defs {
    Some(column_defs) => {
        column_defs
            .iter()
            .map(|column_def| column_def.name.clone())
            .zip(values.into_iter()),
    None => todo!(),
  }
}
```
I want to merge those iterators to save below duplicated `map + collect` logic
```rs
.map(|(key, value)| {
  let value = JsonValue::try_from(value)?.to_string();

  Ok(format!("\"{key}\": {value}"))
})
.collect::<Result<Vec<_>>>()?
```

Reply:
[iter_enum](https://docs.rs/iter-enum/latest/iter_enum/)

This would be useful.
You can also see some examples in core/ workspace

Reply:
It works like a charm! I didn't know that was what i had done with `enum Rows` long times ago..